index.js

```diff
@@ -453,6 +383,44 @@ class TuyaDevice extends EventEmitter {
     return Promise.resolve(true);
   }
 
+  _packetHandler(packet) {
+    // Response was received, so stop waiting
+    clearTimeout(this._sendTimeout);
+
+    if (packet.commandByte === 0x09) {
+      debug(`Pong from ${this.device.ip}`);
+      return;
+    }
+
+    if (packet.commandByte === 0x07) {
+      debug('Set succeeded.');
+      return;
+    }
+
+    /**
+     * Emitted when data is returned from device.
+     * @event TuyaDevice#data
+     * @property {Object} data received data
+     * @property {Number} commandByte
+     * commandByte of result
+     * (e.g. 7=requested response, 8=proactive update from device)
+     * @property {Number} sequenceN the packet sequence number
+     */
+    this.emit('data', packet.payload, packet.commandByte, packet.sequenceN);
+
+    // Call data resolver for sequence number
+    if (this._resolvers[packet.sequenceN]) {
+      this._resolvers[packet.sequenceN](packet.payload);
+
+      // Remove resolver
+      delete this._resolvers[packet.sequenceN];
+    } else if (packet.sequenceN === 0) {
+      this._resolvers[Object.keys(this._resolvers)[0]](packet.payload);
```

Comment:
Not sure I get this, without looking into it too much.

Reply:
Added some more comments; hopefully it's clearer now.

Reply:
Forgive me, I'm still trying to follow the logic. What's the significance of the first resolver? What if there are no resolvers ie `Object.keys(this._resolvers)[0]` is `undefined`?

Reply:
> What's the significance of the first resolver?

There is none; but when TuyAPI receives a response with the sequence number set to 0 it needs to resolve to *something*.

> What if there are no resolvers ie Object.keys(this._resolvers)[0] is undefined?

Good point, not sure what should happen.  That shouldn't occur during normal operation but it probably would in some edge case.

Reply:
Do you mind if I take a day or two to wrap my head around this? I'll have a little time to experiment with this during the weekend.

Reply:
Sure, no rush.

Reply:
Oops, sorry "a day or two" turned into so much more! Let's get back to it, shall we?

I took some time to think about this, and I don't think this logic is quite right.

If you issue a set command (0x07) you'll get an acknowledge response (0x07) with the same sequence number, then subsequently a status update (0x08) with sequence 0.

So this means with each set command, this logic eats two resolvers, the second one being whatever is next, not necessarily related.

What we really want is for the Promise to resolve once for each sequence number. In the event that we don't have a matching sequence number in our resolvers, the Promise has already been resolved via the acknowledge message, or is some other kind of message or garbage data we don't care about.

So I think the correct thing to do here is actually nothing. That is, remove the `else if` block, and I think we're golden.

Reply:
Also, consider that we haven't really lost any information if we don't resolve on status update (0x08) since the `data` event is still fired.

Reply:
I might also change `if (this._resolvers[packet.sequenceN])` to `if (packet.sequenceN in this._resolvers)` to make it explicit that we are checking if the member exists, not whether the member is truthy. Tiny detail though.

Reply:
I noticed that we're also adding a Promise for the heartbeat pings (at `this._resolvers[0]`), but never resolve them due to the logic in packet handler that filters them out.

Perhaps allow the heartbeat packet to resolve 0, and move the heartbeat response handling to `_sendPing` to match the logic of `get` and `set`.

Reply:
> So I think the correct thing to do here is actually nothing. That is, remove the else if block, and I think we're golden.

Makes sense now that you explained it like that.

> Also, consider that we haven't really lost any information if we don't resolve on status update (0x08) since the data event is still fired.

True, but if the user is using a Promise/callback style, they don't get the result back.  So I think we should always resolve if the response sequence number has been mapped to a resolver.

> I noticed that we're also adding a Promise for the heartbeat pings (at `this._resolvers[0]`), but never resolve them due to the logic in packet handler that filters them out.

I think `await`ing `this._send()` on line 254 is sufficent for handling the response.

Good catch, though, I added something to delete the heartbeat resolver in the packet handler.

Let me know what you think.


Tests against stub are not currently passing; I'll take a closer look tonight when I have access to a physical device.